### PR TITLE
`reconfig` support for CH Keeper

### LINKED ISSUE #45886: Keeper Dynamic Reconfiguration
ZooKeeper supports the feature called Dynamic Reconfiguration - https://zookeeper.apache.org/doc/r3.8.0/zookeeperReconfig.html It allows to dynamically change cluster configuration through API without manually changing config files. It would be great to support this feature in Keeper too.

### PR DESCRIPTION
**Pull request by myrrc**
Resolves: https://github.com/ClickHouse/ClickHouse/issues/45886
Depends on: https://github.com/eBay/NuRaft/pull/432, https://github.com/ClickHouse/NuRaft/pull/63
Related: https://github.com/ClickHouse/ClickHouse/issues/50251

### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

Support ZooKeeper `reconfig` command for CH Keeper with incremental reconfiguration which can be enabled via `keeper_server.enable_reconfiguration` setting.
Support adding servers, removing servers, and changing server priorities.

### Documentation entry for user-facing changes

Docs PR: https://github.com/ClickHouse/clickhouse-docs/pull/1164

- [x] Documentation is written (mandatory for new features)

### CODE REVIEW

**Comment by myrrc** [comment]
Closes: https://github.com/ClickHouse/ClickHouse/issues/45886

**Comment by robot-ch-test-poll** [comment]
This is an automated comment for commit b0c462693293e78981a63fd2da6de96ed7ef1827 with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/49450/b0c462693293e78981a63fd2da6de96ed7ef1827/ci_running.html)
The overall status of the commit is 🟡 pending

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🟡 pending</td></tr>
<tr><td>Docs Check</td><td>Builds and tests the documentation</td><td>🟡 pending</td></tr>
</table>

File: src/Coordination/NuraftServer.h [hunk 1]
```diff
@@ -0,0 +1,47 @@
+#pragma once
+#include <Core/Types.h>
+#include <fmt/core.h>
+
+namespace DB
+{
+// nuraft::srv_config is neither copy nor move constructible
+struct NuraftServer { int id; String endpoint; bool learner{false}; int priority{1}; };
```
Comment (alesapin) [inline src/Coordination/NuraftServer.h:8]: Too general name. Maybe `NuraftServerDescription`?

File: src/Coordination/NuraftServer.h [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+using NuraftServers = std::vector<NuraftServer>;
+
+struct AddServer : NuraftServer {};
```
Comment (alesapin) [inline src/Coordination/NuraftServer.h:11]: Inheritance looks very strange here. Use composition please

File: src/Coordination/NuraftServer.cpp [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+    splitInto<':', '='>(parts, server);
+
+    const bool id_endpoint = parts.size() == 3;
+    const bool with_server_type = parts.size() == 4;
+    const bool with_priority = parts.size() == 5;
+    if (!id_endpoint && !with_server_type && !with_priority) return std::nullopt;
+
+    const std::string_view id = parts[0];
+    if (!id.starts_with("server.")) return std::nullopt;
+    if (std::from_chars(&id[7], id.end(), out.id).ec != std::error_code{}) return std::nullopt;
```
Comment (alesapin) [inline src/Coordination/NuraftServer.cpp:22]: I cannot understand this

File: src/Coordination/KeeperStorage.cpp [hunk 4]
```diff
@@ -1344,6 +1351,202 @@ struct KeeperStorageSetRequestProcessor final : public KeeperStorageRequestProce
...
+    auto response_ptr = zk_request->makeResponse();
+
+    if (const auto result = storage.commit(zxid); result != Coordination::Error::ZOK)
+    {
+        response_ptr->error = result;
+        return response_ptr;
+    }
+
+    const KeeperStorage::Node& node = storage.container.getValue(keeper_config_path);
+    decltype(auto) response = dynamic_cast<Coordination::ZooKeeperReconfigResponse &>(*response_ptr);
```
Comment (alesapin) [inline src/Coordination/KeeperStorage.cpp:1541]: please rewrite with type.

File: src/Coordination/KeeperServer.cpp [hunk 5]
```diff
@@ -734,173 +741,38 @@ std::vector<int64_t> KeeperServer::getDeadSessions()
     return state_machine->getDeadSessions();
 }
 
-ConfigUpdateActions KeeperServer::getConfigurationDiff(const Poco::Util::AbstractConfiguration & config)
+bool KeeperServer::tryApplyConfigurationUpdate(const ConfigUpdateAction & task)
```
Comment (alesapin) [inline src/Coordination/KeeperServer.cpp:744]: This function looks unused.

File: src/Coordination/KeeperServer.cpp [hunk 6]
```diff
@@ -734,173 +741,38 @@ std::vector<int64_t> KeeperServer::getDeadSessions()
...
-    else if (task.action_type == ConfigUpdateActionType::RemoveServer)
+    else if (const auto* remove_server = std::get_if<RemoveServer>(&task))
     {
-        LOG_INFO(log, "Will try to remove server with id {}", task.server->get_id());
-
-        bool removed = false;
-        if (task.server->get_id() == state_manager->server_id())
+        if (isLeader() && remove_server->id == state_manager->server_id())
         {
-            LOG_INFO(
```
Comment (alesapin) [inline src/Coordination/KeeperServer.cpp:801]: Why all log messages removed?

File: src/Coordination/KeeperDispatcher.h [hunk 7]
```diff
@@ -89,7 +89,9 @@ class KeeperDispatcher
     void sessionCleanerTask();
     /// Thread create snapshots in the background
     void snapshotThread();
-    /// Thread apply or wait configuration changes from leader
+
+    // TODO: this should be removed once "reconfig" is stabilized
+    void updateConfigurationWithReconfigDisabledThread();
```
Comment (alesapin) [inline src/Coordination/KeeperDispatcher.h:94]: Where is the definition?

File: base/base/find_symbols.h [hunk 8]
```diff
@@ -447,7 +447,7 @@ inline char * find_last_not_symbols_or_null(char * begin, char * end)
 /// See https://github.com/boostorg/algorithm/issues/63
 /// And https://bugs.llvm.org/show_bug.cgi?id=41141
 template <char... symbols, typename To>
-inline void splitInto(To & to, const std::string & what, bool token_compress = false)
+inline To& splitInto(To & to, std::string_view what, bool token_compress = false)
```
Comment (alesapin) [inline base/base/find_symbols.h:450]: Please move to separate PR

File: src/Common/ConcurrentBoundedQueue.h [hunk 9]
```diff
@@ -62,7 +62,7 @@ class ConcurrentBoundedQueue
         return true;
     }
 
-    template <bool front>
+    template <bool front, bool copy = false>
     bool popImpl(T & x, std::optional<UInt64> timeout_milliseconds)
```
Comment (alesapin) [inline src/Common/ConcurrentBoundedQueue.h:66]: Flag completely changes the semantic of function, maybe rename popOrCopyImpl?

**Review by alesapin — COMMENTED** [review]
@antonio2368 , please also take a look when it will be finished.

File: src/Coordination/KeeperDispatcher.cpp [hunk 10]
```diff
@@ -736,6 +732,49 @@ void KeeperDispatcher::updateConfigurationThread()
...
+    using namespace std::chrono_literals;
+    while (!shutdown_called)
+    {
+        ConfigUpdateAction action;
+        if (!update_configuration_queue.copyFront(action))
+            return;
+
+        if (server->applyConfigurationUpdate(action))
+        {
+            (void)update_configuration_queue.tryPop(action);
```
[the remark this answers is not in the document]
Comment (myrrc) [reply src/Coordination/KeeperDispatcher.cpp:746]: The idea is, we need to peek the front element of the queue
(as if it's rejected, we don't lose it).
I guess it can also be implemented with pop + pushFront, will reconsider

File: src/Coordination/KeeperStorage.cpp [hunk 11]
```diff
@@ -1344,6 +1362,212 @@ struct KeeperStorageSetRequestProcessor final : public KeeperStorageRequestProce
...
+}
+
+std::vector<KeeperStorage::Delta> KeeperStorageReconfigRequestProcessor::preprocess(
+    KeeperStorage & storage, int64_t zxid, int64_t, int64_t time, uint64_t & digest, const KeeperContext & ctx) const
+{
+    using enum Coordination::Error;
+    ProfileEvents::increment(ProfileEvents::KeeperReconfigRequest);
+
+    if (!ctx.dispatcher->reconfigEnabled())
+        return {{zxid, ZUNIMPLEMENTED}};
```
Comment (myrrc) [inline src/Coordination/KeeperStorage.cpp:1479]: ZK C client has a special return code for that https://github.com/search?q=repo%3Aapache%2Fzookeeper%20zreconfigdisabled&type=code, although server just throws a NonImplemented error

**Comment by myrrc** [comment]
Summary of discussion with @antonio2368 
1. Having a znode as a regular node in KeeperStorage brings [inconsistencies](https://github.com/ClickHouse/ClickHouse/pull/49450/files#diff-dbf5606f605734e2f5d6335548b5f91e728fb9084255e3d397dee333ac792679R107). Thus it's better not to use it as all needed information is already in KeeperStateMachine cluster config.
2. We can process `reconfig` requests separately from others and leave KeeperStorage untouched at the cost of returning dummy `znodestat`. This is fine as users would get cluster reconfiguration anyway. If we don't use KeeperStorage, we don't get weird things while replaying history for new/stale replicas.

File: src/Coordination/KeeperStateMachine.cpp [hunk 12]
```diff
@@ -291,14 +295,91 @@ bool KeeperStateMachine::preprocess(const KeeperStorage::RequestForSession & req
     return true;
 }
 
+KeeperStorage::ResponseForSession KeeperStateMachine::processReconfiguration(
+    const KeeperStorage::RequestForSession& request_for_session)
+{
+    ProfileEvents::increment(ProfileEvents::KeeperReconfigRequest);
```
Comment (antonio2368) [inline src/Coordination/KeeperStateMachine.cpp:301]: Okay, the reconfig is still forwarded to `app_log`.
what I had in mind was to catch it in KeeperDispatcher in `requestThread` and redirect it to `processReconfiguration` there.

With current approach we still have a problem with replays. Currently you check:
```
    if (!dispatcher.clusterUpdateQueueEmpty())
        return bad_request(ZRECONFIGINPROGRESS);
```

But this can produce different results depending on many different factors like network health (first time we apply `add 2`, server 2 joins instantly and the second reconfigure is processed). After restart there is slower network and it takes time for server 2 to join so the second reconfigur is skipped with `ZRECONFIGINPROGRESS`.

Also, when a node becomes a leader, it search in logs for `conf` log with largest `log_idx` and stores it again, this can undo or create many inconsistent states.

I propose 2 options for reconfig result data:
- instantly return when we push the diff to queue and return the current cluster serialized WITHOUT the diff applied. User can do `get` request on that node to see if the update was applied.
- a bit more complex, but there is `NewConfig` callback in NuRaft that can be used in `KeeperServer` in which we push to the response queue the new cluster serialized OR even better override `commit_conf`. This can break some guarantees from ZK (it will not be ordered with other writes) but I don't think we should care that much about it.

I prefer the first option as it's simpler and makes more sense for async approach. In both solutions we lose persistence slightly but it's much safer and the difference in persistence is minimal (instantly appending vs confirming first that the node is available before appending)

File: src/Coordination/RaftServerConfig.cpp [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+    const bool with_server_type = parts.size() == 3;
+    const bool with_priority = parts.size() == 4;
+    if (!with_id_endpoint && !with_server_type && !with_priority)
+        return std::nullopt;
+
+    const std::string_view id_str = parts[0];
+    if (!id_str.starts_with("server."))
+        return std::nullopt;
+
+    int id;
```
Comment (antonio2368) [inline src/Coordination/RaftServerConfig.cpp:34]: use fixed size integers, e.g. `int32_t` or `Int32`

File: src/Coordination/KeeperServer.h [hunk 14]
```diff
@@ -122,17 +128,12 @@ class KeeperServer
...
-    /// Synchronously check for update results with retries.
-    void applyConfigurationUpdate(const ConfigUpdateAction & task);
-
+    bool applyConfigUpdate(const ClusterUpdateAction& action);
 
-    /// Wait configuration update for action. Used by followers.
-    /// Return true if update was successfully received.
-    bool waitConfigurationUpdate(const ConfigUpdateAction & task);
+    // TODO (myrrc) these functions should be removed once "reconfig" is stabilized
+    void applyConfigUpdateWithReconfigDisabled(const ClusterUpdateAction& action);
```
[the remark this answers is not in the document]
Comment (myrrc) [reply src/Coordination/KeeperServer.h:132]: My idea is that eventually there will be the only option to update cluster via "reconfig" command, so this is more or less a deprecation-like message

File: src/Coordination/KeeperStateMachine.cpp (hunk 12 above)
Comment (myrrc) [reply src/Coordination/KeeperStateMachine.cpp:314]: Ok, makes sense. Will use the first approach

**Comment by myrrc** [comment]
An issue with removing command from the app log is the inability to remove leader from a follower, thus tests replace_leader* fail

File: src/Coordination/KeeperDispatcher.cpp [hunk 15]
```diff
@@ -88,6 +90,12 @@ void KeeperDispatcher::requestThread()
                 if (shutdown_called)
                     break;
 
+                if (request.request->getOpNum() == Coordination::OpNum::Reconfig)
```
Comment (antonio2368) [inline src/Coordination/KeeperDispatcher.cpp:93]: we also pop requests below in line 114

File: src/Coordination/KeeperStateMachine.cpp [hunk 16]
```diff
@@ -291,14 +295,107 @@ bool KeeperStateMachine::preprocess(const KeeperStorage::RequestForSession & req
...
+            response.session_id);
+    }
+}
+
+KeeperStorage::ResponseForSession KeeperStateMachine::processReconfiguration(
+    const KeeperStorage::RequestForSession& request_for_session)
+{
+    ProfileEvents::increment(ProfileEvents::KeeperReconfigRequest);
+
+    const auto& request = static_cast<const Coordination::ZooKeeperReconfigRequest&>(*request_for_session.request);
```
Comment (antonio2368) [inline src/Coordination/KeeperStateMachine.cpp:316]:
```suggestion
    const auto & request = static_cast<const Coordination::ZooKeeperReconfigRequest&>(*request_for_session.request);
```

File: utils/keeper-data-dumper/main.cpp [hunk 17]
```diff
@@ -15,7 +15,7 @@ using namespace DB;
 
 void dumpMachine(std::shared_ptr<KeeperStateMachine> machine)
 {
-    auto & storage = machine->getStorage();
+    auto & storage = machine->getStorageForUnitTests();
```
Comment (antonio2368) [inline utils/keeper-data-dumper/main.cpp:18]: Please revert the name to `getStorage` because it looks a bit strange calling `ForUnitTests` when it's not unit tests.

**Review by antonio2368 — COMMENTED** [review]
This looks good in general, I just need to take a closer look at the tests

File: utils/keeper-data-dumper/main.cpp (hunk 17 above)
Comment (myrrc) [reply utils/keeper-data-dumper/main.cpp:18]: Ok, but we need sort of a note this is not for basic usage as it violates invariants (or we need a lock in this method).
I'll change to the latter.

Comment (myrrc) [reply utils/keeper-data-dumper/main.cpp:18]: What do you think about renaming to getStorageUnsafe?
We can't really use TSA as we return a reference without the ability to attach a lock to it

File: tests/integration/test_keeper_reconfig_add/configs/keeper2.xml [hunk 18]
```diff
@@ -0,0 +1,21 @@
...
+
+        <coordination_settings>
+            <operation_timeout_ms>5000</operation_timeout_ms>
+            <session_timeout_ms>10000</session_timeout_ms>
+            <raft_logs_level>trace</raft_logs_level>
+        </coordination_settings>
+
+        <raft_configuration>
+            <server> <id>1</id> <hostname>node1</hostname> <port>9234</port> </server>
+            <server> <id>2</id> <hostname>node2</hostname> <port>9234</port> </server>
```
Comment (antonio2368) [inline tests/integration/test_keeper_reconfig_add/configs/keeper2.xml:18]: all configs can start with only that node defined
e.g. for node 2
```xml
   <raft_configuration>
            <server> <id>2</id> <hostname>node2</hostname> <port>9234</port> </server>
    </raft_configuration
```

for node 3
```xml
   <raft_configuration>
            <server> <id>3</id> <hostname>node3</hostname> <port>9234</port> </server>
    </raft_configuration
```

Comment (myrrc) [reply tests/integration/test_keeper_reconfig_add/configs/keeper2.xml:18]: No, they can't, as they form a single node cluster and everything messes up. As is Zookeeper, let's assume we start new nodes with new cluster config, not in single mode

File: src/Coordination/KeeperStateMachine.cpp [hunk 19]
```diff
@@ -291,14 +295,107 @@ bool KeeperStateMachine::preprocess(const KeeperStorage::RequestForSession & req
...
+    };
+
+    if (!storage->checkACL(keeper_config_path, Coordination::ACL::Write, session_id, true))
+        return bad_request(ZNOAUTH);
+
+    KeeperDispatcher& dispatcher = *keeper_context->getDispatcher();
+    if (!dispatcher.reconfigEnabled())
+        return bad_request(ZUNIMPLEMENTED);
+    if (!dispatcher.clusterUpdateQueueEmpty())
+        return bad_request(ZRECONFIGINPROGRESS);
```
[the remark this answers is not in the document]
Comment (myrrc) [reply src/Coordination/KeeperStateMachine.cpp:337]: Assuming we can't check for multiple reconfigs if requests land on different nodes, it's better to remove this check at all

File: src/Coordination/KeeperDispatcher.cpp [hunk 20]
```diff
@@ -745,27 +750,72 @@ void KeeperDispatcher::updateConfigurationThread()
...
 
+void KeeperDispatcher::clusterUpdateThread()
+{
+    while (!shutdown_called)
+    {
+        ClusterUpdateAction action;
+        if (!cluster_update_queue.pop(action))
+            return;
+
+        if (server->applyConfigUpdate(action))
```
[the remark this answers is not in the document]
Comment (myrrc) [reply src/Coordination/KeeperDispatcher.cpp:761]: The idea of cancelling reconfigs looks neat, I'll definitely implement it as a separate PR

[the remark this answers is not in the document]
Comment (myrrc) [reply src/Coordination/KeeperDispatcher.cpp:761]: Retrying wouldn't protect us from passing an invalid host or so, as it would be handled after Raft's add_srv returns.

File: src/Coordination/KeeperDispatcher.h [hunk 21]
```diff
@@ -132,10 +134,9 @@ class KeeperDispatcher
     /// and achieved quorum
     bool isServerActive() const;
 
-    /// Registered in ConfigReloader callback. Add new configuration changes to
-    /// update_configuration_queue. Keeper Dispatcher apply them asynchronously.
-    /// 'macros' are used to substitute macros in endpoint of disks
     void updateConfiguration(const Poco::Util::AbstractConfiguration & config, const MultiVersion<Macros>::Version & macros);
+    void pushClusterUpdates(ClusterUpdateActions&& actions);
```
Comment (antonio2368) [inline src/Coordination/KeeperDispatcher.h:138]:
```suggestion
    void pushClusterUpdates(ClusterUpdateActions && actions);
```

File: src/Coordination/KeeperReconfiguration.cpp [hunk 22]
```diff
@@ -0,0 +1,91 @@
...
+}
+
+ClusterUpdateActions leavingToClusterUpdates(const ClusterConfigPtr & cfg, std::string_view leaving)
+{
+    std::vector<std::string_view> leaving_arr;
+    splitInto<','>(leaving_arr, leaving);
+    if (leaving_arr.size() >= cfg->get_servers().size())
+        return {};
+
+    std::unordered_set<int> remove_ids;
```
Comment (antonio2368) [inline src/Coordination/KeeperReconfiguration.cpp:42]:
```suggestion
    std::unordered_set<int32_t> remove_ids;
```

File: src/Coordination/KeeperReconfiguration.cpp [hunk 23]
```diff
@@ -0,0 +1,91 @@
...
+        remove_ids.emplace(id);
+    }
+
+    return out;
+}
+
+String serializeClusterConfig(const ClusterConfigPtr & cfg, const ClusterUpdateActions & updates)
+{
+    RaftServers new_config;
+    std::unordered_set<int> remove_update_ids;
```
Comment (antonio2368) [inline src/Coordination/KeeperReconfiguration.cpp:68]:
```suggestion
    std::unordered_set<int32_t> remove_update_ids;
```

File: src/Coordination/KeeperServer.cpp [hunk 24]
```diff
@@ -783,9 +789,45 @@ std::vector<int64_t> KeeperServer::getDeadSessions()
     return state_machine->getDeadSessions();
 }
 
-ConfigUpdateActions KeeperServer::getConfigurationDiff(const Poco::Util::AbstractConfiguration & config)
+bool KeeperServer::applyConfigUpdate(const ClusterUpdateAction& action)
```
Comment (antonio2368) [inline src/Coordination/KeeperServer.cpp:792]:
```suggestion
bool KeeperServer::applyConfigUpdate(const ClusterUpdateAction & action)
```

File: src/Coordination/RaftServerConfig.cpp [hunk 25]
```diff
@@ -0,0 +1,95 @@
...
+    if (out.priority < 0)
+        return std::nullopt;
+
+    return out;
+}
+
+RaftServers parseRaftServers(std::string_view servers)
+{
+    std::vector<std::string_view> server_arr;
+    std::unordered_set<int> ids;
```
Comment (antonio2368) [inline src/Coordination/RaftServerConfig.cpp:72]:
```suggestion
    std::unordered_set<int32_t> ids;
```

File: src/Coordination/RaftServerConfig.cpp [hunk 26]
```diff
@@ -0,0 +1,95 @@
...
+            const int id = maybe_server->id;
+            if (ids.contains(id))
+                return {};
+
+            out.emplace_back(std::move(*maybe_server));
+            endpoints.emplace(std::move(endpoint));
+            ids.emplace(id);
+        }
+        else
+            return {};
```
Comment (antonio2368) [inline src/Coordination/RaftServerConfig.cpp:91]:
```suggestion
    for (auto & server : splitInto<','>(server_arr, servers))
    {
        if (auto maybe_server = RaftServerConfig::parse(server))
        {
            String endpoint = maybe_server->endpoint;
            if (endpoints.contains(endpoint))
                return {};
            const int id = maybe_server->id;
            if (ids.contains(id))
                return {};

            out.emplace_back(std::move(*maybe_server));
            endpoints.emplace(std::move(endpoint));
            ids.emplace(id);
        }
        else
            return {};
    }
```

**Review by antonio2368 — APPROVED** [review]
